### PR TITLE
Add uint16 images support for co-occurrence matrix

### PR DESCRIPTION
**Pull request by ThomasWalter**
## Description
Now, co-occurrence matrices can be calculated for 16 bit images, too. 

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/): only few modifications in the same style as the original. 
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py): I modified the docs in place. No additional function added
- [ ] Gallery example in `./doc/examples` (new features only): not applicable
- [x] Unit tests

## References
It closes issue #2080 .

### CODE REVIEW

File: skimage/feature/_texture.pyx [hunk 1]
```diff
@@ -13,7 +13,7 @@ cdef extern from "numpy/npy_math.h":
     double NAN "NPY_NAN"
 
 
-def _glcm_loop(cnp.uint8_t[:, ::1] image, double[:] distances,
+def _glcm_loop(cnp.uint16_t[:, ::1] image, double[:] distances,
```
Comment (soupault) [inline skimage/feature/_texture.pyx]: Perform the corresponding change in the docstring. At the moment, it is `Input image, which is converted to the _uint8_ data type.`.

File: skimage/feature/_texture.pyx [hunk 2]
```diff
@@ -29,7 +29,10 @@ def _glcm_loop(cnp.uint8_t[:, ::1] image, double[:] distances,
     levels : int
         The input image should contain integers in [0, levels-1],
         where levels indicate the number of grey-levels counted
-        (typically 256 for an 8-bit image)
+        (typically 256 for an 8-bit image). Be aware that the co-occurrence
+        matrix for every angle and every distance is of size levels x levels. 
+        Choosing a too large level might result in exceedingly large matrix.
+        If you have 16 bit or 32 bit images, consider binning.
```
Comment (soupault) [inline skimage/feature/_texture.pyx]: I'd not duplicate the details here. Please, consider keeping the docstring in .pyx short.

File: skimage/feature/texture.py [hunk 3]
```diff
@@ -30,8 +30,10 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
     levels : int, optional
         The input image should contain integers in [0, levels-1],
         where levels indicate the number of grey-levels counted
-        (typically 256 for an 8-bit image). The maximum value is
-        256.
+        (typically 256 for an 8-bit image). Be aware that the co-occurrence
+        matrix for every angle and every distance is of size levels x levels. 
+        Choosing a too large level might result in exceedingly large matrix.
+        If you have 16 bit or 32 bit images, consider binning.
```
Comment (soupault) [inline skimage/feature/texture.py]: Change the docstring to represent the new behaviour:
```
image : array_like of uint8
    Integer typed input image. The image will be cast to uint8, so the maximum value must be less than 256.
```

File: skimage/feature/texture.py [hunk 4]
```diff
@@ -97,11 +99,21 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
     assert_nD(distances, 1, 'distances')
     assert_nD(angles, 1, 'angles')
 
-    assert levels <= 256
     image = np.ascontiguousarray(image)
     assert image.min() >= 0
-    assert image.max() < levels
-    image = image.astype(np.uint8)
+
+    image_max = image.max()
```
Comment (soupault) [inline skimage/feature/texture.py]: Is there any reason why to use an adaptive `levels`? I'd suggest to make it 256 for uint8 and 65536 for uint16. This way it will also be possible to avoid API change.

File: skimage/feature/texture.py [hunk 5]
```diff
@@ -97,11 +99,21 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
...
+    if levels is None:
+        # if levels is not given, we assume that there are [0, image_max -1] levels to 
+        # be considered. There would be only zeros for all other rows and columns.
```
Comment (soupault) [inline skimage/feature/texture.py]: Please, refactor the comments here and below: make them impersonal, fix syntax, etc.

**Comment by soupault** [comment]
@ThomasWalter thanks for contributing!
This PR needs some work to be done. Please, check my comments on the changes.

File: skimage/feature/texture.py (hunk 4 above)
Comment (ThomasWalter) [inline skimage/feature/texture.py]: Thanks for all the comments. 

I think it is actually better to have `levels` adaptive. The function generates a `levels x levels` output matrix (for each distance and each angle). If the maximum value << `levels`, there will many zeros. An alternative would be to use sparse matrices. In any case, as a user I would not expect that the co-occurrence matrix covers the entire range defined by the data type unless I wish to combine the co-occurrence matrices of many images, which seems a bit exotic as an application (and which is still feasible, even if `levels` is adaptive. 

But if there is a strong incentive to have the matrices cover the entire data range, I can of course change this accordingly. Using the max just seems more logical to me.

**Comment by ThomasWalter** [comment]
Besides the comments you mentioned, is there something else to do? 
In particular, the AppVeyor built failed. Has this something to do with the modifications to this function??

File: skimage/feature/texture.py (hunk 4 above)
Comment (soupault) [inline skimage/feature/texture.py]: From my point of view, the output has to be either full 2**{8/16} x 2**{8/16} matrix of coefficients, or sparse matrix/dictionary with non-zero coefficients only.
We can implement both alternatives and add a parameter for user to choose between them. The default alternative has to be the same as the current one. Otherwise, it will be an API change and we have to follow deprecation path.

> If the maximum value << levels, there will many zeros.
Also, the might be many zeros within the [0:levels] range. As for me, clipping from the upper bound of a range doesn't seem to be the _right_ simplification.

By the way, for uint16, if have set all things correctly, there will be `65536**2 * 2 bytes` matrix, what equals to ~8GB. Not suitable.

**Comment by soupault** [comment]
@ThomasWalter don't worry about the status of AppVeyor build. Let's focus on the substantial side.

I think we have to agree on the form of GLCM we'd like to return first.
After that, we'd need to implement some test for this enhancement and, finally, add your name to the list of contributors.

File: skimage/feature/texture.py (hunk 4 above)
Comment (ThomasWalter) [inline skimage/feature/texture.py]: > The default alternative has to be the same as the current one. Otherwise, it will be an API change and we have to follow deprecation path.
I get your point. 

> Also, the might be many zeros within the [0:levels] range. As for me, clipping from the upper bound of a range doesn't seem to be the right simplification.
You are right in principle, but in practice there are many cases where the full range is not used (i.e. most levels are used in [0, max], but max << max(datatype)). 

> By the way, for uint16, if have set all things correctly, there will be 65536**2 * 2 bytes matrix, what equals to ~8GB. Not suitable.
I know. This was the reason why I changed the default, so that the function can still support some 16bit images (those that do not use a close a large data range). 

 As I mentioned before, my concrete application was to use the function on label images in order to retrieve all touching regions. There are two options: I write a dedicated function to retrieve the touching regions (I do not know whether skimage would be interested in this). Or I give back a sparse matrix. This should solve the problem for all reasonably sized 16 bit images.

Comment (soupault) [inline skimage/feature/texture.py]: > You are right in principle, but in practice there are many cases where the full range is not used (i.e. most levels are used in [0, max], but max << max(datatype)).

We have parameter `levels` for this, don't we?

> 2e32 x 2e32

Yes, sorry, the total size of a full GLCM for uin16 will be, at least, twice bigger, because we have uint32 for counters, so `>= (2**16)*(2**16) * 4 / 1024**3 = 16 (GB)`.

---

OK, I can see two ways to go, at the moment:

1. We allow the function to work on full uint16 range __only__ if parameter `labels` is specified. If it is not, we return an error with detailed reasoning (or warning, actually);
2. We add parameter `return_sparse` (or something like this) to the function, implement sparse GLCM computation and allow uint16 images to be used only with `return_sparse=True`.

In both alternatives, the default `labels` value is 2**img.dtype, i.e. full range.

We can get some inspiration from MATLAB (http://www.mathworks.com/help/images/ref/graycomatrix.html). They explicitly clip (`GrayLimits`) and bin (`NumLevels`). Defaults are reasonable (`If I is an intensity image, graycomatrix scales the image to eight gray-levels.`) as they declare even uint64 images support.

Comment (ThomasWalter) [inline skimage/feature/texture.py]: OK, sorry I could not work on this anymore the last weeks. 

So, I chose option 1. We cannot really imitate Matlab for this anymore, because the parameter `levels` does unfortunately not match neither of the two parameters `GrayLimits` and `NumLevels`. If I wanted to match the behavior of `NumLevels`, the function would really change w.r.t what it does now. 

I make a new pull request ...

File: skimage/feature/texture.py [hunk 6]
```diff
@@ -11,27 +11,26 @@
                        _multiblock_lbp)
 
 
-def greycomatrix(image, distances, angles, levels=256, symmetric=False,
+def greycomatrix(image, distances, angles, levels=None, symmetric=False,
                  normed=False):
-    """Calculate the grey-level co-occurrence matrix.
+    """ *** YOYO *** Calculate the grey-level co-occurrence matrix.
```
Comment (jni) [inline skimage/feature/texture.py]: Um....

?

LOL

File: skimage/feature/texture.py [hunk 7]
```diff
@@ -97,11 +97,31 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
...
-    assert levels <= 256
     image = np.ascontiguousarray(image)
     assert image.min() >= 0
-    assert image.max() < levels
-    image = image.astype(np.uint8)
+
+    image_max = image.max()
+
+    #if image.dtype in [np.float, np.float16, np.float32, np.float]
+    if image.dtype == np.float:
```
Comment (jni) [inline skimage/feature/texture.py]: `if np.issubdtype(image.dtype, np.float)`

File: skimage/feature/texture.py [hunk 8]
```diff
@@ -97,11 +97,31 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
...
+        raise ValueError("Float images are not supported by greycomatrix."
```
Comment (jni) [inline skimage/feature/texture.py]: Add space after period for when this string is concatenated with below:

```python
>>> raise ValueError('hello'
... 'world')
```

```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-4a130ed1e39b> in <module>()
----> 1 raise ValueError('hello'
      2 'world')

ValueError: helloworld
```

File: skimage/feature/texture.py [hunk 9]
```diff
@@ -97,11 +97,31 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
...
+                         "The image needs to be casted to an unsigned integer type.")
```
Comment (jni) [inline skimage/feature/texture.py]: English: casted -> cast.

File: skimage/feature/texture.py (hunk 7 above)
Comment (soupault) [inline skimage/feature/texture.py]: @jni whoa, that's neat! Does numpy have this function for the lowest version we support, though?

File: skimage/feature/texture.py [hunk 10]
```diff
@@ -97,11 +97,31 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
...
+    
+    # for 16 bit images, levels must be set.
+    if image.dtype != np.uint8:
+        if levels is None:
+            raise ValueError("For images other than 8 bit images, argument levels must be set to a user defined value."
+                             "A large value will produce a large output matrix, potentially causing memory problems."
+                             "In this case, one might consider binning the image prior to calling greycomatrix.")
```
Comment (jni) [inline skimage/feature/texture.py]: All code must fit in 80 characters. This message probably needs to be made a bit more concise, e.g.:

"The `levels` argument is required for data types other than uint8. "
"The resulting matrix will be at least `levels ** 2` in size."

If you want to prescribe the solution to users, I would suggest doing so in a note in the docstring, rather than in the exception.

File: skimage/feature/texture.py [hunk 11]
```diff
@@ -97,11 +97,31 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
...
+        
+    if levels is None:    
+        levels = 256
+
+    assert image_max < levels
```
Comment (jni) [inline skimage/feature/texture.py]: Add a failure message to this assert.

File: skimage/feature/texture.py [hunk 12]
```diff
@@ -97,11 +97,31 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
...
+
+    # image is cast to uint16 (because of fixed typing in cython) 
+    image = image.astype(np.uint16)
```
Comment (jni) [inline skimage/feature/texture.py]: So, I'm concerned about the performance implications of this... Could you look at [fused types](http://docs.cython.org/src/userguide/fusedtypes.html) as the solution?

**Comment by jni** [comment]
@ThomasWalter thanks very much for this! I hate restricting data types so I'm very pleased about this. Could you address my comments? Let me know if you have questions/concerns!

File: skimage/feature/texture.py (hunk 6 above)
Comment (ThomasWalter) [inline skimage/feature/texture.py]: OK, that's embarrassing ...

File: skimage/feature/texture.py (hunk 12 above)
Comment (ThomasWalter) [inline skimage/feature/texture.py]: Excellent. This is exactly what I was looking for in the first place. I will try to go for this. Thx.

**Comment by ThomasWalter** [comment]
Ok I guess that I have addressed all comments; thanks a lot for this. Maybe the doc string is still a bit wordy ... you tell me.

File: skimage/feature/_texture.pyx [hunk 13]
```diff
@@ -12,24 +12,38 @@ from .._shared.transform cimport integrate
 cdef extern from "numpy/npy_math.h":
     double NAN "NPY_NAN"
 
-
-def _glcm_loop(cnp.uint8_t[:, ::1] image, double[:] distances,
+ctypedef fused uint_8_16:
```
Comment (jni) [inline skimage/feature/_texture.pyx]: I guess I would call this type "any_int". =)

File: skimage/feature/_texture.pyx [hunk 14]
```diff
@@ -12,24 +12,38 @@ from .._shared.transform cimport integrate
...
                double[:] angles, Py_ssize_t levels,
                cnp.uint32_t[:, :, :, ::1] out):
     """Perform co-occurrence matrix accumulation.
 
     Parameters
     ----------
     image : ndarray
-        Input image, which is converted to the uint8 data type.
+        Integer typed input image. Only positive valued images are supported. 
+        If type is uint16, the argument `levels` needs to be set.
```
Comment (jni) [inline skimage/feature/_texture.pyx]: "If image type is other than uint8", since this comment now applies also to 32-bit, 64-bit, etc.

File: skimage/feature/_texture.pyx [hunk 15]
```diff
@@ -12,24 +12,38 @@ from .._shared.transform cimport integrate
...
         List of pixel pair distance offsets.
     angles : ndarray
         List of pixel pair angles in radians.
     levels : int
-        The input image should contain integers in [0, levels-1],
+        The input image should contain integers in [0, `levels`-1],
         where levels indicate the number of grey-levels counted
-        (typically 256 for an 8-bit image)
+        (typically 256 for an 8-bit image). This argument is required for
+        16-bit images and is typically the maximum of the image.
```
Comment (jni) [inline skimage/feature/_texture.pyx]: 16-bit or higher

File: skimage/feature/texture.py [hunk 16]
```diff
@@ -20,18 +20,21 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
 
     Parameters
     ----------
-    image : array_like of uint8
-        Integer typed input image. The image will be cast to uint8, so
-        the maximum value must be less than 256.
+    image : array_like of uint8 or uint16.
+        Integer typed input image. Only positive valued images are supported. 
+        If type is uint16, the argument `levels` needs to be set.
```
Comment (jni) [inline skimage/feature/texture.py]: As above

**Comment by jni** [comment]
Thanks Thomas! just a few more minor comments...!

File: skimage/feature/texture.py (hunk 16 above)
Comment (ThomasWalter) [inline skimage/feature/texture.py]: Ok. Thanks a lot for your comments. I modified the doc strings accordingly. 

One question: do I also need to include new tests?

**Comment by jni** [comment]
Excellent, thank you! :+1: from me! @scikit-image/core we need another reviewer on this.

File: skimage/feature/_texture.pyx [hunk 17]
```diff
@@ -12,24 +12,38 @@ from .._shared.transform cimport integrate
...
-        Input image, which is converted to the uint8 data type.
+        Integer typed input image. Only positive valued images are supported. 
+        If type is other than uint8, the argument `levels` needs to be set.
     distances : ndarray
         List of pixel pair distance offsets.
     angles : ndarray
         List of pixel pair angles in radians.
     levels : int
-        The input image should contain integers in [0, levels-1],
+        The input image should contain integers in [0, `levels`-1],
```
Comment (stefanv) [inline skimage/feature/_texture.pyx]: So, we support signed types but not signed values?

Comment (jni) [inline skimage/feature/_texture.pyx]: This is correct and intuitive, since you can't index with negative values. (One could imagine an elaborate scheme with negative indices, but that is not really worth thinking about!) The signed type support is handy because I imagine that many people unwittingly end up with int64 images.

Comment (stefanv) [inline skimage/feature/_texture.pyx]: Hrm, but GLCM only count when it sees identical values paired up.  So those
values *can* be negative.

File: skimage/feature/_texture.pyx [hunk 18]
```diff
@@ -12,24 +12,38 @@ from .._shared.transform cimport integrate
 cdef extern from "numpy/npy_math.h":
     double NAN "NPY_NAN"
 
-
-def _glcm_loop(cnp.uint8_t[:, ::1] image, double[:] distances,
+ctypedef fused any_int:
```
Comment (stefanv) [inline skimage/feature/_texture.pyx]: Would cython.integral suffice here?

See cython/docs/build/html/src/userguide/fusedtypes.html (sorry, the Cython docs are down right now, so I cloned the repo and found the documentation there).

File: skimage/feature/texture.py (hunk 4 above)
Comment (stefanv) [inline skimage/feature/texture.py]: Another suggestion: return a dense array, but allow "levels" be specified as an array (or be ``array.ravel().unique()``, e.g.).

**Comment by stefanv** [comment]
Please also add a test for the newly added type support.

File: skimage/feature/_texture.pyx (hunk 17 above)
Comment (ThomasWalter) [inline skimage/feature/_texture.pyx]: It is possible to calculate a GLCM for negative valued images, but the indices of the resulting matrix would not correspond to the grey levels anymore. Then we would need to give back the map indices vs. grey levels (which would change the API). Or we always give back the full data range matrix, starting with the negative values. In practice however, a full data range matrix for types > uint8 does not fit in the RAM of most computers, so this is not a good solution neither. 

If we want to do this properly, we could do binning prior to GLCM calculation (see discussion with soupault). But would this still be inline with the current API? I have the impression that this would change the function quite dramatically. Binning makes a lot of sense for this function, but it can also be left to the user.

File: skimage/feature/texture.py (hunk 4 above)
Comment (ThomasWalter) [inline skimage/feature/texture.py]: Sounds good. So the GLCM will be len(levels) x len(levels), if levels is array like, and [0, levels - 1] if levels is a positive integer. This I could do. 

However,  `levels` cannot default to `array.ravel().unique()`, as this would result in different matrices for different images. For most applications, this is not the desired behavior. Also, it would change the default behavior, which is probably also not what we want.

File: skimage/feature/_texture.pyx (hunk 18 above)
Comment (ThomasWalter) [inline skimage/feature/_texture.pyx]: I just tested, and apparently this is not sufficient. I did not find information on whether unsigned are also part of python.integral.

**Comment by ThomasWalter** [comment]
OK, I added a test function. It only tests different image types (but not different values), and I answered to the other comments. 

I guess that we need to decide what we want to do for the `levels` parameter, and in particular whether we would allow for an array. On the upside, this would allow us to include binning, and it would allow us to return a dense array even for types > uint8. On the downside, the resulting matrix would be less handy in a way, as the indices do not correspond to grey levels anymore. Also the default, that if `levels` is a number, then the GLCM is calculated for the levels `[0, levels-1]` seems not so elegant. 

What would you think?

File: skimage/feature/texture.py (hunk 4 above)
Comment (jni) [inline skimage/feature/texture.py]: I think we should leave these enhancements for a separate PR.

**Comment by jni** [comment]
@ThomasWalter I don't think it's unintuitive to use a scalar `levels` as a way to get `[0, levels - 1]` as a range. It coincides with virtually all scientific Python indexing! =) To make it more intuitive: `levels` means levels in the range `[0, levels)`. ;)

I think adding an array version can wait for a new PR.

File: skimage/feature/texture.py [hunk 19]
```diff
@@ -97,11 +101,27 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
...
+
+    image_max = image.max()
+
+    if np.issubdtype(image.dtype, np.float):
+        raise ValueError("Float images are not supported by greycomatrix. "
+                         "The image needs to be cast to an unsigned integer type.")
+    
+    # for 16 bit images (or larger), levels must be set.
+    if image.dtype != np.uint8:
+        if levels is None:
```
Comment (jni) [inline skimage/feature/texture.py]: Collapse this to single clause, makes it easier to parse:

```python
if levels is None and image.dtype != np.uint8:
```

File: skimage/feature/texture.py [hunk 20]
```diff
@@ -97,11 +101,27 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
...
+            raise ValueError("The levels argument is required for data types other than uint8. "
+                             "The resulting matrix will be at least levels ** 2 in size.")
+                             
+        
+    if levels is None:    
+        levels = 256
+
+    assert image_max < levels, "The image maximum needs to be smaller than `levels`."
```
Comment (jni) [inline skimage/feature/texture.py]: Why is this an assert rather than another valueerror?

File: skimage/feature/texture.py [hunk 21]
```diff
@@ -97,11 +101,27 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
...
+        raise ValueError("Float images are not supported by greycomatrix. "
+                         "The image needs to be cast to an unsigned integer type.")
+    
+    # for 16 bit images (or larger), levels must be set.
+    if image.dtype != np.uint8:
+        if levels is None:
+            raise ValueError("The levels argument is required for data types other than uint8. "
+                             "The resulting matrix will be at least levels ** 2 in size.")
+                             
+
```
Comment (jni) [inline skimage/feature/texture.py]: Now that we have decided to support signed integer types, we might want to add another check:

```python
if image.dtype in (np.int8, np.int16, np.int32, np.int64) and np.any(image < 0):
    raise ValueError("greycomatrix is undefined for images with negative values.")
```

Comment (jni) [inline skimage/feature/texture.py]: Additionally, all of these ValueErrors are not currently tested anywhere! Use `assert_raises` to test that the correct error is raised in each circumstance.

**Comment by jni** [comment]
Just a few more comments!

File: skimage/feature/texture.py (hunk 20 above)
Comment (ThomasWalter) [inline skimage/feature/texture.py]: Just because it was like this in the original code. Changed it now.

File: skimage/feature/texture.py (hunk 21 above)
Comment (ThomasWalter) [inline skimage/feature/texture.py]: There has been an assert before: 
`assert image.min() >= 0`
which I change into a ValueError. I guess that replacing `image.min() >= 0` by `np.any(image < 0)` and also checking the type is a bit more efficient, too ... 

I will add the tests now.

**Comment by jni** [comment]
@ThomasWalter Looks good to me! I don't think the Travis failure is related to the PR... @stefanv?

**Comment by stefanv** [comment]
No, the Travis failure on OSX is due to a refactoring of Terryfy by
Matthew, so go ahead and merge.

File: skimage/feature/_texture.pyx [hunk 22]
```diff
@@ -7,29 +7,44 @@ cimport numpy as cnp
...
                double[:] angles, Py_ssize_t levels,
                cnp.uint32_t[:, :, :, ::1] out):
     """Perform co-occurrence matrix accumulation.
 
     Parameters
     ----------
     image : ndarray
-        Input image, which is converted to the uint8 data type.
+        Integer typed input image. Only positive valued images are supported. 
+        If type is other than uint8, the argument `levels` needs to be set.
```
Comment (soupault) [inline skimage/feature/_texture.pyx]: `levels` has to be always set here. Please, remove this line.

File: skimage/feature/_texture.pyx [hunk 23]
```diff
@@ -7,29 +7,44 @@ cimport numpy as cnp
...
     distances : ndarray
         List of pixel pair distance offsets.
     angles : ndarray
         List of pixel pair angles in radians.
     levels : int
-        The input image should contain integers in [0, levels-1],
+        The input image should contain integers in [0, `levels`-1],
         where levels indicate the number of grey-levels counted
-        (typically 256 for an 8-bit image)
+        (typically 256 for an 8-bit image). This argument is required for
```
Comment (soupault) [inline skimage/feature/_texture.pyx]: Same here. Please, reformulate.

File: skimage/feature/texture.py [hunk 24]
```diff
@@ -97,11 +101,28 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
...
     image = np.ascontiguousarray(image)
-    assert image.min() >= 0
-    assert image.max() < levels
-    image = image.astype(np.uint8)
+
+    image_max = image.max()
+
+    if np.issubdtype(image.dtype, np.float):
+        raise ValueError("Float images are not supported by greycomatrix. "
+                         "The image needs to be cast to an unsigned integer type.")
```
Comment (soupault) [inline skimage/feature/texture.py]: `or signed`? :)

File: skimage/feature/texture.py [hunk 25]
```diff
@@ -97,11 +101,28 @@ def greycomatrix(image, distances, angles, levels=256, symmetric=False,
...
+                         "The resulting matrix will be at least levels ** 2 in size.")
+                             
+    if image.dtype in (np.int8, np.int16, np.int32, np.int64) and np.any(image < 0):
+        raise ValueError("Negative valued images are not supported.")
+
+    if levels is None:    
+        levels = 256
+
+    if image_max >= levels:
+        raise ValueError("The image maximum needs to be smaller than `levels`.")
```
Comment (soupault) [inline skimage/feature/texture.py]: This was hard to get. Maybe put it in the other way: "`levels` needs to be ... than the image maximum"?

File: skimage/feature/texture.py (hunk 24 above)
Comment (jni) [inline skimage/feature/texture.py]: Not signed. Signed support is in as backup, but since we don't support negative image values, we should leave this as unsigned so that it's unambiguous what needs to happen.

**Comment by jni** [comment]
@soupault since those are minor issues, I'm going to go ahead and merge (since @stefanv also gave the go-ahead) and fix them in a separate PR.

Thank you @ThomasWalter!

File: skimage/feature/_texture.pyx (hunk 22 above)
Comment (jni) [inline skimage/feature/_texture.pyx]: @soupault this is incorrect, see line 123

Comment (soupault) [inline skimage/feature/_texture.pyx]: @jni Sorry? I mean this statement is not correct for _cython_ function.
For the one in `skimage/feature/texture.py` it is obviously so.
